src/approaches/grammar_search_invention_approach.py

```diff
@@ -18,6 +18,91 @@
 from predicators.src.settings import CFG
 
 
+################################################################################
+#                          Programmatic classifiers                            #
+################################################################################
+
+class _ProgrammaticClassifier(abc.ABC):
+    """A classifier implemented an arbitrary program.
+    """
+    @abc.abstractmethod
+    def __call__(self, s: State, o: Sequence[Object]) -> bool:
+        """All programmatic classifiers are functions of state and objects.
+
+        The objects are the predicate arguments.
+        """
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def __str__(self) -> str:
+        raise NotImplementedError("Override me!")
+
+
+class _NullaryClassifier(_ProgrammaticClassifier):
+    """A classifier on zero objects.
+    """
+    def __call__(self, s: State, o: Sequence[Object]) -> bool:
+        assert len(o) == 0
+        return self._classify_state(s)
+
+    @abc.abstractmethod
+    def _classify_state(self, s: State) -> bool:
+        raise NotImplementedError("Override me!")
+
+
+class _UnaryClassifier(_ProgrammaticClassifier):
+    """A classifier on one object.
+    """
+    def __call__(self, s: State, o: Sequence[Object]) -> bool:
+        assert len(o) == 1
+        return self._classify_object(s, o[0])
+
+    @abc.abstractmethod
+    def _classify_object(self, s: State, obj: Object) -> bool:
+        raise NotImplementedError("Override me!")
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _SingleAttributeCompareClassifier(_UnaryClassifier):
+    """Compare a single feature value with a constant value.
+    """
+    object_index: int
+    attribute_name: str
+    constant: float
+    compare: Callable[[float, float], bool]
+    compare_str: str
+
+    def _classify_object(self, s: State, obj: Object) -> bool:
+        return self.compare(s.get(obj, self.attribute_name), self.constant)
+
+    def __str__(self) -> str:
+        return f"({self.object_index}.{self.attribute_name}" + \
+               f"{self.compare_str}{self.constant:.3})"
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class _ForallClassifier(_NullaryClassifier):
+    """Apply a predicate to all objects.
```

Comment:
Would a learned Handempty() predicate basically be, "for all objects, their holding bit in the low-level state is false"?

Reply:
yeah